src/execute.rs

```diff
@@ -0,0 +1,654 @@
+) -> Result<Response, ContractError> {
+    let sender = info.sender.to_string();
+
+    // Get the pool the position will be created in
+    let pool: Pool = PoolmanagerQuerier::new(&deps.querier)
+        .pool(pool_id)?
+        .pool
+        .ok_or(ContractError::PoolNotFound { pool_id: pool_id })?
+        .try_into()
+        .unwrap();
+    let spread_factor = Decimal256::from_str(pool.spread_factor.as_str())?;
+
+    // Get a rough estimate of the ratio of assets in the pool at the tick range the user wants to create a position in
+    let (asset_0_ratio, asset_1_ratio) =
+        calc_asset_ratio_from_ticks(upper_tick, pool.current_tick, lower_tick)?;
+
+    deps.api.debug(&format!(
+        "initial asset ratios: asset_0_ratio {asset_0_ratio:?}, asset_1_ratio {asset_1_ratio:?}"
+    ));
+
+    let token_out_denom: String;
+    let token_in: Coin;
+    let token_provided_swap_amount: Uint128;
+
+    // Depending on which asset we are swapping in, utilize the ratio of assets calculated above to determine
+    // how much of the the tokenIn we provided will be used for swapping
+    if token_provided.denom == pool.token0.clone() {
+        if pool.current_tick > upper_tick {
+            // If the current tick is greater than the upper tick, we will swap the entire amount of token0 for token1
+            token_provided_swap_amount = token_provided.amount;
+        } else {
+            // Otherwise, we utilize the ratio of assets calculated above to determine how much of token0 to swap for token1
+            let token_provided_dec =
+                Decimal256::from_str(token_provided.amount.to_string().as_str())?;
+            token_provided_swap_amount = token_provided_dec
+                .checked_mul(asset_1_ratio.checked_add(spread_factor)?)?
+                .to_uint_floor()
+                .try_into()?;
+        }
+        token_out_denom = pool.token1.clone();
+        token_in = Coin {
+            denom: pool.token0.clone(),
+            amount: token_provided_swap_amount,
+        };
+    } else if token_provided.denom == pool.token1.clone() {
+        if pool.current_tick < lower_tick {
+            // If the current tick is less than the lower tick, we will swap the entire amount of token1 for token0
+            token_provided_swap_amount = token_provided.amount;
+        } else {
+            // Otherwise, we utilize the ratio of assets calculated above to determine how much of token1 to swap for token0
+            let token_provided_dec =
+                Decimal256::from_str(token_provided.amount.to_string().as_str())?;
+            token_provided_swap_amount = token_provided_dec
+                .checked_mul(asset_0_ratio.checked_add(spread_factor)?)?
+                .to_uint_floor()
+                .try_into()?;
+        }
+        token_out_denom = pool.token0.clone();
+        token_in = Coin {
+            denom: pool.token1.clone(),
+            amount: token_provided_swap_amount,
+        };
```

Comment:
nit: we could probably reduce code duplication here by extracting a helper. Something like

```rust
pub fn helper(all_consumedl: bool, asset_ratio: type) -> ...
        if all_consumed{
          
            token_provided_swap_amount = token_provided.amount;
        } else {
for token0
            let token_provided_dec =
                Decimal256::from_str(token_provided.amount.to_string().as_str())?;
            token_provided_swap_amount = token_provided_dec
                .checked_mul(asset_ratio.checked_add(spread_factor)?)?
                .to_uint_floor()
                .try_into()?;
        }
        token_out_denom = pool.token0.clone();
        token_in = Coin {
            denom: pool.token1.clone(),
            amount: token_provided_swap_amount,
        };

        // return
}
```